core/src/main/java/org/streamprocessor/core/helpers/FailsafeElement.java

```diff
@@ -0,0 +1,66 @@
+package org.streamprocessor.core.helpers;
+
+import java.io.Serializable;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.Getter;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.hamcrest.core.IsInstanceOf;
+import org.json.JSONObject;
+import org.streamprocessor.core.io.bigquery.FailureFields;
+
+@Getter
+public class FailsafeElement<OriginalT, CurrentT> implements Serializable {
+    private final OriginalT originalElement;
+    @Nullable private final CurrentT currentElement;
+    private String pipelineStep;
+    private String exceptionType;
+    private Throwable exceptionDetails;
+    private String eventTimestamp;
+
+    public FailsafeElement(OriginalT originalElement, @Nullable CurrentT currentElement) {
+        this.originalElement = originalElement;
+        this.currentElement = currentElement;
+    }
+
+    public static <OriginalT, CurrentT> FailsafeElement<OriginalT, CurrentT> of(
+            OriginalT originalElement, CurrentT currentElement) {
+        return new FailsafeElement<>(originalElement, currentElement);
+    }
+
+    public FailsafeElement<OriginalT, CurrentT> setPipelineStep(String pipelineStep) {
+        this.pipelineStep = pipelineStep;
+        return this;
+    }
+
+    public FailsafeElement<OriginalT, CurrentT> setExceptionType(String exception) {
+        this.exceptionType = exception;
+        return this;
+    }
+
+    public FailsafeElement<OriginalT, CurrentT> setExceptionDetails(Throwable exceptionDetails) {
+        this.exceptionDetails = exceptionDetails;
+        return this;
+    }
+
+    public FailsafeElement<OriginalT, CurrentT> setEventTimestamp(String eventTimestamp) {
+        this.eventTimestamp = eventTimestamp;
+        return this;
+    }
+
+    // TODO: WIP
+    public <U extends PubsubMessage> PubsubMessage getPubsubMessage() {
```

Comment:
Looks good I think! Why do we have to extend the pubsub class? I guess we can just bundle the data however we like as a json object and put it inside a pubsub message?

Payload draft:
{"originalPayload":
"pipelineStep":
"exceptionType"
etc
}

And then create a pubsub message with attributes and return it?

Reply:
Something like this?
        return new PubsubMessage(payload.toString().getBytes(), ((PubsubMessage) originalElement).getAttributeMap());

Reply:
> Why do we have to extend the pubsub class?

Trying to ensure that the method could only be used if the `OriginalT` type is a PubsubMessage, I was hoping that this would give an error directly in the IDE but it did not

Reply:
I added this code instead to make it clear
```java
if (!(originalElement instanceof PubsubMessage)) {
            throw new IllegalArgumentException("Original element is not of type PubsubMessage.");
        }
```